tracer/src/Datadog.Trace/DiagnosticListeners/AspNetCoreDiagnosticObserver.cs

```diff
@@ -290,25 +290,27 @@ private static void SetLegacyResourceNames(BeforeActionStruct typedArg, Span spa
                 routeTemplate = $"{controllerName}/{actionName}";
             }
 
-            string resourceName = $"{httpMethod} {routeTemplate}";
-
-            // override the parent's resource name with the MVC route template
-            span.ResourceName = resourceName;
+            return $"{httpMethod} {routeTemplate}";
         }
 
-        private static Span StartMvcCoreSpan(Tracer tracer, Span parentSpan, BeforeActionStruct typedArg, HttpContext httpContext, HttpRequest request)
+        private static Span StartMvcCoreSpan(
+            Tracer tracer,
+            AspNetCoreHttpRequestHandler.RequestTrackingFeature trackingFeature,
+            BeforeActionStruct typedArg,
+            HttpContext httpContext,
+            HttpRequest request)
         {
             // Create a child span for the MVC action
             var mvcSpanTags = new AspNetCoreMvcTags();
-            var mvcScope = tracer.StartActiveInternal(MvcOperationName, parentSpan.Context, tags: mvcSpanTags);
+            var mvcScope = tracer.StartActiveInternal(MvcOperationName, tags: mvcSpanTags);
             tracer.TracerManager.Telemetry.IntegrationGeneratedSpan(IntegrationId);
             var span = mvcScope.Span;
             span.Type = SpanTypes.Web;
 
-            // This is only called with new route names, so parent tags are always AspNetCoreEndpointTags
-            var parentTags = (AspNetCoreEndpointTags)parentSpan.Tags;
```

Comment:
This was where the cast was failing before - it assumed the parentSpan was our aspnetcore span, which it wasn't necessarily!